src/util/hash.rs

```diff
@@ -29,58 +29,55 @@ use consensus::encode::Encodable;
+pub fn bitcoin_merkle_root_inline<T>(hashes: &mut [T]) -> Option<T>
     where T: Hash + Encodable,
           <T as Hash>::Engine: io::Write,
 {
-    if hashes.len() == 0 {
-        panic!("merkle root of an empty tree is undefined");
+    match hashes.len() {
+        0 => None,
+        1 => Some(hashes[0]),
+        _ => Some(merkle_root_r(hashes)),
     }
-
-    // If there is only a single hash then that hash is, by definition, the merkle root.
-    if hashes.len() == 1 {
-        return hashes[0];
-    }
-
-    merkle_root_r(hashes)
 }
 
 /// Calculates the merkle root of an iterator of hashes.
 ///
-/// # Panics
-/// The merkle root of an empty tree of hashes is undefined, panics if `hashes` is empty.
-pub fn bitcoin_merkle_root<T, I>(mut hashes: I) -> T
+/// # Returns
+/// - `None` if `hashes` is empty. The merkle root of an empty tree of hashes is undefined.
+/// - `Some(hash)` if `hashes` contains one element. A single hash is by definition the merkle root.
+/// - `Some(merkle_root)` if length of `hashes` is greater than one.
+pub fn bitcoin_merkle_root<T, I>(mut hashes: I) -> Option<T>
     where T: Hash + Encodable,
           <T as Hash>::Engine: io::Write,
           I: ExactSizeIterator<Item = T>,
 {
-    if hashes.len() == 0 {
-        panic!("merkle root of an empty tree is undefined");
-    }
-
-    // If there is only a single hash then that hash is, by definition, the merkle root.
-    if hashes.len() == 1 {
-        return hashes.next().expect("cannot fail, length == 1");
-    }
-
-    // We need a local copy to pass to `merkle_root_r`. It's more efficient to do the first loop of
-    // processing as we make the copy instead of copying the whole iterator.
-    let half_len = hashes.len() / 2 + hashes.len() % 2;
-    let mut alloc = Vec::with_capacity(half_len);
-    while let Some(hash1) = hashes.next() {
-        // If the size is odd, use the last element twice.
-        let hash2 = hashes.next().unwrap_or(hash1);
-        let mut encoder = T::engine();
-        hash1.consensus_encode(&mut encoder).unwrap();
-        hash2.consensus_encode(&mut encoder).unwrap();
-        alloc.push(T::from_engine(encoder));
+    match hashes.len() {
+        0 => None,
+        1 => Some(hashes.next().expect("cannot fail, length == 1")),
+        _ => {
+            // We need a local copy to pass to `merkle_root_r`. It's more efficient to do the first loop of
+            // processing as we make the copy instead of copying the whole iterator.
+            let half_len = hashes.len() / 2 + hashes.len() % 2;
+            let mut alloc = Vec::with_capacity(half_len);
+            while let Some(hash1) = hashes.next() {
+                // If the size is odd, use the last element twice.
+                let hash2 = hashes.next().unwrap_or(hash1);
+                let mut encoder = T::engine();
+                hash1.consensus_encode(&mut encoder).unwrap();
+                hash2.consensus_encode(&mut encoder).unwrap();
```

Comment:
nit: two `unwrap`s which may be replaced with `expect`

Reply:
I left these on purpose so they could be done separately, you commented on that PR anyways but linking here for others

https://github.com/rust-bitcoin/rust-bitcoin/pull/719